Synonym Mgmnt API:  PUT request

carlosdelest opened the pull request:
Moves synonym set system index code to the server module.

Adds an initial PUT action that uses multiple documents to store synonym sets internally. See [here](https://docs.google.com/document/d/1ZIJER4qEab-OacZuPMYRP57g-T45ZkuCvqOnJSCBlUs/edit#heading=h.busv0z6buwmp) for detailed context on index structure and design decision.

`PUT /_synonyms/<synonyms_set_id>`

```json
{
    "synonyms_set": [
        {
            "id": "32rfs",
            "synonyms": "hello, hi"
        },
        {          
            "synonyms": "bye, goodbye"
        },
        {          
            "synonyms": "test => check"
        }
    ]
}
```

Response:

```json
{
    "result": "updated"
}
```

In case the synonym set is created, `"result": "updated"` will be returned.

It uses a single doc to store the synonym set. This will be changed, kept simple for now to demonstrate the capabilities and validate the approach.

Commit:
Add first implementation of PUT API

Commit:
Synonyms as arrays of strings

Commit:
Move Synonyms to server

Commit:
Remove base package from module

carlosdelest commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/CommonAnalysisPlugin.java:
```diff
@@ -149,7 +146,8 @@
 import static org.elasticsearch.plugins.AnalysisPlugin.requiresAnalysisSettings;
 
-public class CommonAnalysisPlugin extends Plugin implements AnalysisPlugin, ScriptPlugin, SystemIndexPlugin {
+public class CommonAnalysisPlugin extends Plugin implements AnalysisPlugin, ScriptPlugin {
```
Removed system index creation from `CommonAnalysisPlugin`

carlosdelest commented on server/src/main/java/org/elasticsearch/action/ActionModule.java:
```diff
@@ -434,6 +436,7 @@
 import org.elasticsearch.rest.action.search.RestMultiSearchAction;
 import org.elasticsearch.rest.action.search.RestSearchAction;
 import org.elasticsearch.rest.action.search.RestSearchScrollAction;
+import org.elasticsearch.rest.action.synonyms.RestPutSynonymsAction;
```
Registered a new PUT action. Please check that the different actions are in the expected packages for server 🙏

mayya-sharipova replied:
This package is fine, or we can also put this action into `org.elasticsearch.rest.action.admin.indices` as it is about .synonyms index

carlosdelest commented on server/src/main/java/org/elasticsearch/synonyms/SynonymSet.java:
```diff
@@ -0,0 +1,67 @@
+public class SynonymSet implements Writeable, ToXContentObject {
+
+    public static final ParseField SYNONYMS_FIELD = new ParseField("synonyms");
+    private static final ConstructingObjectParser<SynonymSet, Void> PARSER = new ConstructingObjectParser<>("synonyms", args -> {
```
For now, synonym set is specified as an array of strings - this was the original approach and I decided to keep it for simplicity and change it later.

Commit:
Fix test

mayya-sharipova commented on server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymsAction.java:
```diff
@@ -0,0 +1,130 @@
+        }
+    }
+
+    public static class Response extends ActionResponse implements StatusToXContentObject {
```
We also need to expose `_version` field in the response. It is especially important for a single document design as I would expect a document will undergo a lot of modifications (and may be concurrent modifications).

Which makes me wonder if we should make `version` parameter as compulsory in a request?

carlosdelest replied:
I'll add it, thanks!

I don't think it should be mandatory - let it be the user's choice depending on the use case for updating synonyms. Maybe some are not expecting concurrent updates, or update from a single process...

The UI will definitely benefit from using it

Commit:
Check FF before adding action handlers

Commit:
Synonym index has one doc per synonym rule

Commit:
Response includes created or updated status

Commit:
Add tests for request and response serialization

Commit:
Merge remote-tracking branch 'origin/main' into carlosdelest/synonyms-move-to-server

# Conflicts:
#	modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/CommonAnalysisPluginTests.java

Commit:
Extract Synonyms API FF to its own class

Commit:
Add first YAML test, including feature flags

Commit:
Fix YAML test

Commit:
Fixing YAML tests

carlosdelest commented on server/src/main/java/org/elasticsearch/action/ActionModule.java:
```diff
@@ -767,6 +771,11 @@ public <Request extends ActionRequest, Response extends ActionResponse> void reg
         actions.register(FetchHealthInfoCacheAction.INSTANCE, FetchHealthInfoCacheAction.TransportAction.class);
 
+        // Synonyms
+        if (SynonymsAPI.isEnabled()) {
```
A new SynonymsAPI class for doing the FF check has been created - it didn't seem right to depend on the whole SynonymsManagementService to do just that

mayya-sharipova replied:
I am not clear what dependency you talking about. As `SynonymsManagementService` is already part of `server`, we don't introduce any extra depedency?

mayya-sharipova replied:
Thanks, Carlos, makes sense to remove this class when FF is removed.

carlosdelest commented on server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:
```diff
@@ -0,0 +1,164 @@
+        // TODO Add synonym rules validation
+
+        // Delete synonym set if it existed previously
+        DeleteByQueryRequest dbqRequest = new DeleteByQueryRequest(SYNONYMS_INDEX).setQuery(
```
PUT request first deletes all synonym docs using a delete by query, and then uses a bulk insert

mayya-sharipova replied:
What's the reason for using `delete by query` request?  Why do we need to all documents belonging to the this synonymSetID on update? I thought the whole point of using multiple documents is that we don't need to do this. 

I think what we should do instead is to do a bulk update request:
- create SynonymRules that don't have IDs
- update SynonymRules that have IDs

carlosdelest replied:
The PUT update will replace the whole synonym set, so it makes sense to me to first remove all existing docs. Keep in mind that the IDs might not have been included from the user, and I think it's good to let the user change the whole synonym set without needing to include the rules IDs.

> I thought the whole point of using multiple documents per synonymSetID is to avoid this, and work on individual document base.

Correct! But that would be done at the individual synonym set level. So it will be possible to do:

`PUT /_synonyms/<synonym_set>/<rule_id> `

to override a single rule. But using

`PUT /_synonyms/<synonym_set>`

will replace the whole synonym set, effectively removing all synonym set rules and adding them from the request.

Does it make sense?

carlosdelest commented on x-pack/qa/core-rest-tests-with-security/build.gradle:
```diff
@@ -5,6 +5,7 @@ dependencies {
   clusterModules project(':modules:mapper-extras')
   clusterModules project(':modules:rank-eval')
   clusterModules project(':modules:ingest-common')
+  clusterModules project(':modules:reindex')
```
Reindex module needs to be added as it contains delete_by_query impl

elasticsearchmachine commented:
Hi @carlosdelest, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

carlosdelest committed:
Update docs/changelog/95895.yaml

mayya-sharipova commented on server/src/main/java/org/elasticsearch/synonyms/SynonymsAPI.java:
```diff
@@ -0,0 +1,19 @@
+
+import org.elasticsearch.common.util.FeatureFlag;
+
+public class SynonymsAPI {
```
I wonder if this class needed only to store `SYNONYMS_API_FEATURE_FLAG`?

 If yes, why can't we keep `SYNONYMS_API_FEATURE_FLAG` in `SynonymsManagementAPIService`?

carlosdelest replied:
Answered above - seemed cleaner to export just that FF as part of a class, instead of using the whole `SynonymsManagementAPIService`, that exports more operations.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:
```diff
@@ -0,0 +1,164 @@
+                    {
+                        builder.startObject("synonym");
+                        {
+                            builder.field("type", "text");
```
may be use `match_only_text`, as I don't think we will ever need scoring.

carlosdelest replied:
I'm ok with that. Will using `match_only_text` prevent us from using "prefix", "wildcard", or "regexp" queries though? That might be useful for providing a search experience...

We might as well choose not to pre-optimize on this and decide later...

mayya-sharipova replied:
No, it doesn't; you can do "prefix", "wildcard", or "regexp" queries with `match_only_text` as well.

mayya-sharipova replied:
@carlosdelest What is the decision on this comment, should we use `text` or `match_only_text` should be enough?

carlosdelest replied:
I've changed it to `match_only_text`. Thanks for the discussion!

mayya-sharipova commented on server/src/main/java/org/elasticsearch/indices/SystemIndices.java:
```diff
@@ -116,7 +118,8 @@ public class SystemIndices {
     private static final Map<String, Feature> SERVER_SYSTEM_FEATURE_DESCRIPTORS = Stream.of(
-        new Feature(TASKS_FEATURE_NAME, "Manages task results", List.of(TASKS_DESCRIPTOR))
+        new Feature(TASKS_FEATURE_NAME, "Manages task results", List.of(TASKS_DESCRIPTOR)),
+        new Feature(SYNONYMS_FEATURE_NAME, "Manages synonyms", List.of(SYNONYMS_DESCRIPTOR))
```
Should we add .synonyms index only `SYNONYMS_API_FEATURE_FLAG` is enabled?

carlosdelest replied:
Good catch! Adding that 👍

mayya-sharipova commented:
@carlosdelest  Thanks for the PR, I took an initial look, and left some comments. Some other comments:
1. Should we update the PR description to reflect current design.
2. When I try to execute PUT `_synonyms/set1` request in my client with security enabled by defaul, I am getting the following error message:

```java
[2023-05-18T15:20:58,725][WARN ][o.e.h.AbstractHttpServerTransport] [runTask-0] handling request [null][PUT][/_synonyms/set1][Netty4HttpChannel{localAddress=/[0:0:0:0:0:0:0:1]:9200, remoteAddress=/[0:0:0:0:0:0:0:1]:61863}] took [24973ms] which is above the warn threshold of [5000ms]
» [2023-05-18T15:20:58,729][ERROR][o.e.ExceptionsHelper     ] [runTask-0] fatal error java.lang.AssertionError: action.origin [synonyms] is unknown!
»       at org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.ExceptionsHelper.lambda$maybeDieOnAnotherThread$4(ExceptionsHelper.java:257)
»       at java.base/java.util.Optional.ifPresent(Optional.java:178)
»       at org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.ExceptionsHelper.maybeDieOnAnotherThread(ExceptionsHelper.java:247)
»       at org.elasticsearch.transport.netty4@8.9.0-SNAPSHOT/org.elasticsearch.http.netty4.Netty4HttpPipeliningHandler.exceptionCaught(Netty4HttpPipeliningHandler.java:378)
```
To resolve it, I think we need to add synonyms client origin to [uthorizationUtils](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java#L152)

mayya-sharipova commented on server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymsAction.java:
```diff
@@ -0,0 +1,157 @@
+
+        public Request(String name, BytesReference content, XContentType contentType) throws IOException {
+            this.name = name;
+            this.synonymSet = SynonymSet.fromXContent(XContentHelper.createParser(XContentParserConfiguration.EMPTY, content, contentType));
```
Why do we need an extra class `SynonymSet` that just a holder of `List<SynonymRule>`? Can we just read `List<SynonymRule>` here without creating `SynonymSet` object?

carlosdelest replied:
We might want to add information to it in the future - synonyms format, number of total synonym rules. I think it's more extensible this way.

mayya-sharipova replied:
synchronized with @carlosdelest offline, and we decided to keep `SynonymSet` class with an added TODO: to remove it later if not needed.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/action/synonyms/PutSynonymsAction.java:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    public static class Request extends ActionRequest {
+        private final String name;
```
Instead of `name`, should we use  `synonymSetId` or something like this?

carlosdelest replied:
Sounds good, changing!

carlosdelest commented on server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:
```diff
@@ -0,0 +1,164 @@
+                    builder.field("dynamic", "strict");
+                    builder.startObject("properties");
+                    {
+                        builder.startObject("synonym");
```
I wanted to prevent confusion so `synonyms` would refer to the topmost synonym set, and `synonym` to the individual rule:

```
{
    "synonyms": [
        {
            "id": "32rfs",
            "synonym": "hello, hi"
        },
        {          
            "synonym": "bye, goodbye"
        },
        {          
            "synonym": "test => check"
        }
    ]
}
```

Maybe we could use synonym_set for the topmost synonym element, and synonyms for the inner ones:

```
{
    "synonym_set": [
        {
            "id": "32rfs",
            "synonyms": "hello, hi"
        },
        {          
            "synonyms": "bye, goodbye"
        },
        {          
            "synonyms": "test => check"
        }
    ]
}
```

WDYT?

carlosdelest replied:
We got an agreement to use:

```json
{
    "synonyms_set": [
        {
            "id": "32rfs",
            "synonyms": "hello, hi"
        },
        {          
            "synonyms": "bye, goodbye"
        },
        {          
            "synonyms": "test => check"
        }
    ]
}
```

I'll be updating the code with this 👍

Commit:
Add synonym system index only if FF is active

Commit:
PR Review comments, perform update immediately for synonym sets

Commit:
Add security to synonym PUT request

Commit:
Remove origin client, add warnings to YAML tests

Commit:
Merge remote-tracking branch 'carlosdelest/carlosdelest/synonyms-move-to-server' into carlosdelest/synonyms-move-to-server

carlosdelest commented:
> My main [comment](https://github.com/elastic/elasticsearch/pull/95895/files#r1198248280) is how we should approach updating synonym rules. Also about response: the response that we currently getting is a little strange (e.g. even if we create new rules, we get "updated" response)

I think we need to distinguish between a synonym set update and individual synonym rules updates:
- A synonym set update updates all synonym rules, effectively replacing the synonym set. I'd vote for using a simple "update" vs "created" mechanism for this, as I don't think the user is interested in individual updates. In case we want to do that, we should strive for a bulk request style request (which I don't think is part of the requirements).
- An individual synonym rule update would change an individual synonym rule

I think all requirements are met using the current response format for replacing an entire synonym set. We can check with the team if needed, but it is aligned with the current Design doc API spec.

carlosdelest commented:
> Should we update the PR description to reflect current design: detailed description of request and response.
Done, thanks!

> When I try to execute PUT _synonyms/set1 request in my client with security enabled by default, I am getting the following error message:

That was my mistake. I was planning to do security as a separate follow-up. I have removed the origin setting client, and added warning checks for the YAML tests. We can remove the warning checks when the client uses an origin.

Commit:
Spotless fix

Commit:
Add reindex module for delete by query to be available

Commit:
Merge branch 'carlosdelest/synonyms-put-security' into carlosdelest/synonyms-move-to-server

Commit:
Removed allowed warnings code in YAML tests

Commit:
Set origin client in SynonymsManagementAPIService

Commit:
Changed request fields: synonyms_set and synonyms

Commit:
Changes on field names

mayya-sharipova commented on rest-api-spec/src/main/resources/rest-api-spec/api/synonyms.put.json:
```diff
@@ -0,0 +1,39 @@
+  "synonyms.put": {
+    "documentation": {
+      "url": "https://www.elastic.co/guide/en/elasticsearch/reference/master/put-synonyms.html",
+      "description": "Creates or updates a synonym set"
```
Nit: I think we should be consistent everywhere (including API): either "synonym set" or "synonyms set". I am ok with either version (including API).

carlosdelest replied:
Agreed, changing all instances I found in the code 👍

mayya-sharipova commented on server/src/main/java/org/elasticsearch/synonyms/SynonymRule.java:
```diff
@@ -0,0 +1,113 @@
+        return builder;
+    }
+
+    public XContentBuilder toXContentForIndex(XContentBuilder builder, String synonymSetName) throws IOException {
```
This method doesn't seem to be used anywhere. Should we remove it?

carlosdelest replied:
Removed

mayya-sharipova commented on server/src/main/java/org/elasticsearch/synonyms/SynonymsAPI.java:
```diff
@@ -0,0 +1,21 @@
+public class SynonymsAPI {
+    private static final FeatureFlag SYNONYMS_API_FEATURE_FLAG = new FeatureFlag("synonyms_api");
+
+    public static final String SYNONYMS_ORIGIN = "synonyms";
```
If the goal is to remove this class, when we remove the feature flag, should we move `SYNONYMS_ORIGIN` variable to `SynonymsManagementAPIService` class?

carlosdelest replied:
Sounds good - moving

mayya-sharipova commented on server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:
```diff
@@ -0,0 +1,165 @@
+        // Delete synonym set if it existed previously
+        DeleteByQueryRequest dbqRequest = new DeleteByQueryRequest(SYNONYMS_INDEX).setQuery(
+            QueryBuilders.termQuery(SYNONYMS_SET_FIELD, resourceName)
+        ).setIndicesOptions(IndicesOptions.fromOptions(true, true, false, false));
```
Can you please explain why do we need these indices options?

carlosdelest replied:
The idea is not to fail when the system index has not been created yet, as can happen if no synonym has been created.

That way, we don't have to catch the index not found exception, and can use the same code path for it.

I'll document it on the code to make it clear

carlosdelest replied:
I'm afraid it is not created on startup - it's created when the first synonym set is added (checked that). I think we need those index options to simplify error management, but I'm open for suggestions!

mayya-sharipova replied:
Thanks for explanation, @carlosdelest . I had a wrong assumption that the server system indices are created on node startup. Makes sense!

kderusso commented on server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:
```diff
@@ -0,0 +1,165 @@
+    private final Client client;
+
+    public static final SystemIndexDescriptor SYNONYMS_DESCRIPTOR = SystemIndexDescriptor.builder()
+        .setIndexPattern(SYNONYMS_INDEX + "*")
```
Do we need a wildcard index pattern here, since we only have a single index and we're not aliasing multiple indices?

carlosdelest replied:
I guess we're not having versioned indices for this yet, and we can add it later. Changing!

carlosdelest replied:
It seems that yes, we need it - it's checked on [this class](https://github.com/elastic/elasticsearch/blob/a5a645362fb6c8955ba030050e1ecf456b7c6ff4/server/src/main/java/org/elasticsearch/indices/SystemIndices.java#L199) to allow for upgrades

kderusso replied:
Thanks for confirming! Followup question, should we support versioned indices/aliases from the start then?

kderusso commented on server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:
```diff
@@ -0,0 +1,165 @@
+                    listener.onFailure(ex);
+                }
+
+                bulkRequestBuilder.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
```
Is it possible that chaining these two requests could result in a race condition, where you're updating the synonym set with one request and deleting it with another request?

carlosdelest replied:
Yes, it is. Same as with Search Applications approach.

kderusso replied:
That's my concern - perhaps we could find a better path forward here, and apply a similar solution to search applications?

carlosdelest replied:
Same as happened in Search Apps, I don't think this is a blocker as of now. We can think of a better approach in follow up work.

kderusso replied:
Will this be launching as tech preview as well?

mayya-sharipova replied:
My assumption is that this request - update the whole synonyms set  will be very rare, so it is ok if there is race condition.  An alternative to use `synchronize` should slow things a lot, to my mind.

kderusso replied:
We may at least want to create a followup issue to fix.

kderusso commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/SynonymsProfileUser.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * internal user that manages the synonyms index.
+ */
+public class SynonymsProfileUser extends User {
```
Nitpick: Would a better name for this be something like `SynonymsManagerUser` or `SynonymsAdminUser`?

carlosdelest replied:
Changing to SynonymsInternalUser, to make it more clear it's an internal user - though there's not much of a standard here

kderusso reviewed (commented):
Overall changes look great! Left a few questions and comments.

kderusso reviewed (commented):
Overall changes look great! Left a few questions and comments.

carlosdelest commented on server/src/main/java/org/elasticsearch/rest/action/synonyms/RestPutSynonymsAction.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public List<Route> routes() {
+        return List.of(new Route(PUT, "/_synonyms/{name}"));
```
Sounds good, thanks for the catch! 👍

carlosdelest commented on server/src/main/java/org/elasticsearch/synonyms/SynonymRule.java:
```diff
@@ -0,0 +1,113 @@
+        if (Strings.isEmpty(synonym)) {
+            throw new IllegalStateException("synonym cannot be null");
+        }
+        ;
```
It's not, good catch!

Commit:
Code review comments

Commit:
Fix synonyms system index pattern

Commit:
Add synonyms index versioning support

Commit:
Change synonym field type to match_only_text

Commit:
Merge remote-tracking branch 'origin/main' into carlosdelest/synonyms-move-to-server

# Conflicts:
#	x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUsers.java
#	x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/User.java
#	x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java

Commit:
Adapt synonym to InternalUsers, fix alias permission errors

mayya-sharipova commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUsers.java:
```diff
@@ -122,6 +122,21 @@ public class InternalUsers {
         )
     );
 
+    public static final InternalUser SYNONYMS_USER = new InternalUser(
```
Should we provide description (API documentation) similar to other users?

carlosdelest replied:
👍  Done!

mayya-sharipova reviewed (approved):
@carlosdelest Thanks for the great work! This PR lgtm! I've left small comments that doesn't need to me to review again.

Commit:
PR review comments

carlosdelest commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/UsernamesField.java:
```diff
@@ -32,6 +32,8 @@ public final class UsernamesField {
     public static final String ASYNC_SEARCH_ROLE = "_async_search";
     public static final String STORAGE_USER_NAME = "_storage";
     public static final String STORAGE_ROLE_NAME = "_storage";
+    public static final String SYNONYMS_USER_NAME = "_synonyms";
```
Thanks for the catch - I changed `InternalUsers` class so it is used as the synonyms internal user name.

carlosdelest commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/10_synonyms_put.yml:
```diff
@@ -0,0 +1,38 @@
+  - match: { result: "created" }
+
+---
+"Update synonyms set":
```
Sounds good! Done 👍

kderusso reviewed (approved):
Looks good!

cbuescher reviewed (commented):
I did a first round of reviews and left a few smaller questions and one bigger discussion topic around the need of the PUT request to upfront delete previously existing documents.

carlosdelest commented on qa/smoke-test-multinode/build.gradle:
```diff
@@ -16,6 +16,7 @@ restResources {
 dependencies {
   clusterModules project(":modules:mapper-extras")
   clusterModules project(":modules:ingest-common")
+  clusterModules project(":modules:reindex")
```
This is due to using delete by query to remove the rules in the synonym set prior to updating it.

carlosdelest commented on server/src/main/java/org/elasticsearch/synonyms/SynonymsSet.java:
```diff
@@ -0,0 +1,86 @@
+import java.util.List;
+import java.util.Objects;
+
+// TODO review the need for this class, we might use just SynonymRule as this is a just a holder of SynonymRule
```
The follow-up implementation for GET does already use this class directly - let's address this TODO a bit later in the implementation of the CRUD API.

cbuescher replied:
Can we add this class with the follow-up GET then? I'd like to have less open TODOs in the codebase, those tend to not get cleaned up later. Earlier when we developed thing like this on a feature branch this was easier to spot when doing the final merge, but this all goes directly to the main branch. Or remove the TODO altogether when you're sure its going to be used.

carlosdelest replied:
I'd say let's remove this if not needed once we review GET - this is already implemented as is and it's actually used in GET request.

I'd like to change as little as possible on this PR as it's a long running one, it's behind a FF, and there's room for improvement in the near future.

Once we have GET merged we will be able to start including the new loading method in the analyzers and thus start to have an end-to-end feature.

Let's focus on having an end to end, working solution and test the overall viability, and then change the internal implementation and address pending questions if needed be.

carlosdelest commented on server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:
```diff
@@ -0,0 +1,180 @@
+
+        // TODO Add synonym rules validation
+
+        // Delete synonyms set if it existed previously. Avoid catching an index not found error by ignoring unavailable indices
```
PUT should create the element if it does not exist, or replace the element with the new definition in case it already exists ([here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT))

Think of the doc index API: `PUT /test-index/_doc/doc-id` would create or update the doc `doc-id`, and not error out in case it already exists (unless `op_type=create` is specified)

That is the reason why documents are deleted - in case rules exist, they need to be replaced by the new ones.

There will be individual rule updates as well:

PUT /_synonyms/<rule_set_id>/<rule_id>

so we will provide the way to update individual rules (or add new ones) as well. Please see [Individual Synonym Rule Update](https://docs.google.com/document/d/1ZIJER4qEab-OacZuPMYRP57g-T45ZkuCvqOnJSCBlUs/edit#heading=h.79pldhtzq9oq) in the Synonym Management API doc.

carlosdelest commented on docs/changelog/95895.yaml:
```diff
@@ -0,0 +1,5 @@
+pr: 95895
+summary: "Move synonyms set code to server, add PUT request"
+area: Analysis
+type: enhancement
```
Good catch - I guess these are created automatically and we can't skip them.

cbuescher replied:
What happens if we remove the file? If that doesn't work if we remove the "enhancement" label from the PR? I don't like the idea of adding partial work on features with several PRs and every one of them showing up in the changelog. That should be reserved for released features and show a more succinct view of changes for the user.

carlosdelest replied:
I removed the file and added `>refactoring` as PR type, and it worked 👍

mayya-sharipova replied:
@carlosdelest @cbuescher According to our rules, for PRs under feature flags we should use `>non-issue` label.

carlosdelest commented on rest-api-spec/src/main/resources/rest-api-spec/api/synonyms.put.json:
```diff
@@ -0,0 +1,39 @@
+{
+  "synonyms.put": {
+    "documentation": {
+      "url": "https://www.elastic.co/guide/en/elasticsearch/reference/master/put-synonyms.html",
```
Sounds good. I've created a [Synonyms documentation task](https://elasticco.atlassian.net/browse/ES-6180) so we don't forget.

cbuescher commented on server/src/main/java/org/elasticsearch/action/synonyms/TransportPutSynonymsAction.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Override
+    protected void doExecute(Task task, PutSynonymsAction.Request request, ActionListener<PutSynonymsAction.Response> listener) {
+        synonymsManagementAPIService.putSynonymsset(request.synonymsSetId(), request.synonymsset(), listener);
```
Fine, lets revisit if the SynonymsManagementAPIService gets too large at some point. I think that is one of the reasons why in other places we keep the code for performing the actual operations in the transport actions themselves.

Commit:
Remove changelog for non-feature

cbuescher commented on server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:
```diff
@@ -0,0 +1,180 @@
+                                .source(builder);
+                            final String synonymRuleId = synonymRule.id();
+                            if (synonymRuleId != null) {
+                                indexRequest.id(synonymRuleId);
```
If I understand this correctly, using the plain synonym rule id a doc id will potentially overwrite rules using the same id in other synonym sets? I think joining set id / rule Id might be useful as a doc id here.
In that case I also believe the previous delete request isn't necessary because each rule will overwrite any potential predecessor?

carlosdelest replied:
> If I understand this correctly, using the plain synonym rule id a doc id will potentially overwrite rules using the same id in other synonym sets?

That was not the plan - rule set IDs would be isolated from one another. Potentially, two rule sets could contain the same rule ID but the two would be independent.

That's why `ruleset_id` is stored as a separate field in the rule docs - it is used as a filter to retrieve all the rules in a ruleset.

> In that case I also believe the previous delete request isn't necessary because each rule will overwrite any potential predecessor?

The delete request would still be needed, as:
- Users can choose not to specify rule IDs (because they don't care about them, or are not planning to individually update rules)
- Users may "delete" rules by replacing a ruleset, but not specifying some of the rules that were part of the previous one.

carlosdelest replied:
You're right Christoph, thanks for explaining. I was taking advantage of using the ID auto-generation feature from the index request, and didn't thought on potential collisions with user-defined IDs. 

I like your approach to ensure that IDs are unique to each synonym set, by prepending the rule set ID to it.

In case the user does not specify an ID, we could generate one using `synonymSetId + "_" + UUIDs.base64UUID()`. 
 Do you think this approach would work?

cbuescher replied:
Yes, as long as the UI can safely remove that prefix when displaying the ids. 
Is the idea to show auto-generated Ids in error messages and when listing all rules of a set then? Or leave those empty in that case?

carlosdelest replied:
We still haven't thought about that on the UI side. I'd say that ids should not be displayed, as they do not convey useful information to the user (who is using the UI to manage synonyms).

IDs make sense to me from the API perspective, but not from the UI perspective. That is the way synonyms work for App Search today - auto-generated IDs are used for synonyms, but only used on the API side.

carlosdelest commented:
@elasticmachine merge upstream

elasticmachine commented:
merge conflict between base and head

Commit:
Merge remote-tracking branch 'origin/main' into carlosdelest/synonyms-move-to-server

# Conflicts:
#	x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUsers.java

cbuescher commented:
From the issue description:

> It uses a single doc to store the synonym set. This will be changed, kept simple for now to demonstrate the capabilities and validate the approach
I assume this is outdated since it appears several documents are, one for each rule in a rule set?

carlosdelest commented:
> I assume this is outdated since it appears several documents are, one for each rule in a rule set?

🤦 Yes it is. Updating now, sorry for the confusion. This PR has been open since the initial approach, and I didn't check the description when we made the decision on using multiple docs.

Thanks!

Commit:
Merge remote-tracking branch 'origin/main' into carlosdelest/synonyms-move-to-server

Commit:
Generate rules ID prepending synonym set ID

carlosdelest commented:
@cbuescher , thanks for your review! Is there anything blocking from your side? I'm aiming to merge this ASAP so we can work on GET and thus unblock work for Analyzers. Thanks!

cbuescher commented:
Thanks Carlos for the last updates. I think there should be some additional tests in the GET followup that also cover the id overwrite cases mentioned here earlier, but other than that LGTM

carlosdelest commented:
Thanks @cbuescher , will definitely work on the tests! 👍
